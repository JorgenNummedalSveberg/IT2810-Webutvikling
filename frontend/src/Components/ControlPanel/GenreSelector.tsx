import {useDispatch, useSelector} from "react-redux";
import {setGenre} from "../../actions";
import {MenuItem, Select} from "@material-ui/core";
import React from "react";
import {State} from "../../types/State";
import {makeStyles} from "@material-ui/styles";

// Endrer sjanger vi søker etter
function GenreSelector(props: { refresh: () => void }) {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Henter inn sjangre fra redux state
    const genres = useSelector((state: State) => state.genres);

    // Henter inn sjangre fra redux state
    const genreText = useSelector((state: State) => state.filter.genre);
    const genre = {text: genreText, value: genreText};

    // Definerer options for dropdown meny
    const genreOptions = genres.map((genre, index) => {
        return {key: genre, text: genre, value: genre, id: genre};
    });

    // Opdaterer state når du endrer sjanger
    function onSearchChange(e: any, data: any) {
        dispatch(setGenre(data.props.value));
        props.refresh();
    }

    const classes = makeStyles({
        selector: {
            backgroundColor: '#E98074',
        }
    })

    return (
        <Select
            className={classes().selector}
            variant={'outlined'}
            value={genre.value}
            onChange={onSearchChange}
            autoWidth={true}
        >
            {genreOptions.map((genreOption, index) => <MenuItem key={index}
                                                                value={genreOption.value}>{genreOption.text}</MenuItem>)}
        </Select>
    )
}

export default GenreSelector