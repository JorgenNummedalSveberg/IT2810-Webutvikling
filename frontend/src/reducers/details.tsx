// Reducer for lagring av filmen som skal vises i popup
import {Movie} from "../types/Movie";

const movieReducer = (details = {show: false, movie: {}}, action: { type: string; payload: {show: boolean, movie: Movie }; }) => {
    switch (action.type) {
        case 'setMovie':
            return action.payload;
        default:
            return details;
    }
}
export default movieReducer;