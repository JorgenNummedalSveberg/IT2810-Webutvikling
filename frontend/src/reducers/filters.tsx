// Samlet reducer for alle filterene
const filterReducer = (filters = {desc:true, genre: "", search: "", sort: "", score: [0,10], year: [1900,2020]}, action: { type: string; payload: string | boolean | number[]; }) => {
    let result = filters;
    switch (action.type) {
        // Endrer retning vi sorterer i
        case 'setDesc':
            result.desc = action.payload as boolean;
            break;
        // Endrer tittel søkekriterie
        case 'setSearch':
            result.search = action.payload as string;
            break;
        // Endrer sjanger søkekriterie
        case 'setGenre':
            result.genre = action.payload as string;
            break;
        // Endrer hva vi sorterer etter
        case 'setSort':
            result.sort = action.payload as string;
            break;
        // Endrer rating søkekriterie
        case 'setScore':
            result.score = action.payload as number[];
            break;
        // Endrer årstall søkekriterie
        case 'setYears':
            result.year = action.payload as number[];
            break;
    }
    return result;
}
export default filterReducer;