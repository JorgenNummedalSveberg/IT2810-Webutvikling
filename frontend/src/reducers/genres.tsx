// Reducer for Ã¥ endre sjangerlisten, setter en generisk liste som default
const genresReducer = (genres = [
    'Select genre...',
    'Biography', 'Drama',
    'History', 'Crime',
    'Fantasy', 'Mystery',
    'Comedy', 'Romance',
    'Family', 'Sci-Fi',
    'Adventure', 'Action',
    'Thriller', 'War',
    'Horror', 'Western',
    'Animation', 'Musical',
    'Sport', 'Music'
], action: { type: any; payload: any; }) => {
    switch (action.type) {
        case 'setGenres':
            return genres.concat(action.payload).sort();
        default:
            return genres;
    }
}
export default genresReducer;