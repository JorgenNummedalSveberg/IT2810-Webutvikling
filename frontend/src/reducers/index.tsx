import moviesReducer from "./movies";
import genresReducer from "./genres";
import {combineReducers} from "redux";
import filterReducer from "./filters";


// Lager en samlet reducer slik at alt kan lagres i en enkelt store
const reducers = combineReducers({
    movies: moviesReducer,
    genres: genresReducer,
    filter: filterReducer
})

export default reducers;