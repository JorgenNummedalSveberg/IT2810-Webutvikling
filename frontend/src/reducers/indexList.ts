// Reducer for lagring av filmindex som brukes til Ã¥ vite hvilke filmer som skal vises
const indexListReducer = (indexList: string[] = [], action: { type: string; payload: string[]; }) => {
    switch (action.type) {
        case 'indexList':
            return action.payload;
        default:
            return indexList;
    }
}
export default indexListReducer