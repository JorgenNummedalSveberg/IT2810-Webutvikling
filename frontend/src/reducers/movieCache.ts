import {Movie} from "../types/Movie";

// Reducer for lagring av filmer, legger filmer inn i en liste og danner en cache
const movieCacheReducer = (movies: Movie[] = [], action: { type: string; payload: Movie[]; }) => {
    switch (action.type) {
        case 'pushMovies':
            return movies.concat(action.payload);
        default:
            return movies;
    }
}
export default movieCacheReducer