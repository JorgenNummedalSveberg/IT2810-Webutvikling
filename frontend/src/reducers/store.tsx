import {createStore} from "redux";
import reducers from "../reducers";

// Store for redux, @ts-ignore delen er for Ã¥ kunne se redux i redux webtools
export const store = createStore(
    reducers,
    // @ts-ignore
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)