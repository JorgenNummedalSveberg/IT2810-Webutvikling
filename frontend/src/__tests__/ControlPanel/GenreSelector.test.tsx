import React from 'react';
import {render} from '@testing-library/react';
import GenreSelector from "../../Components/ControlPanel/GenreSelector";
import {Provider} from "react-redux";
import {store} from "../../reducers/store";

function mock() {
}

describe("Test to see if the sort button renders with our mock function, and passes props correctly", () => {
    test("Rendering of button", () => {
        const {container} = render(<Provider store={store}><GenreSelector refresh={mock}/></Provider>);
        expect(container).toMatchSnapshot();
    })
})