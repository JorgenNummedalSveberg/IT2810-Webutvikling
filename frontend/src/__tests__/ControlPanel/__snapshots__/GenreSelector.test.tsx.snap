// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to see if the sort button renders with our mock function, and passes props correctly Rendering of button 1`] = `
<div>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root makeStyles-selector-1"
    data-testid="genreSelector"
  >
    <div
      aria-haspopup="listbox"
      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
      role="button"
      tabindex="0"
    >
      Select genre...
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput"
      tabindex="-1"
      value="Select genre..."
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
      style="padding-left: 8px;"
    >
      <legend
        class="PrivateNotchedOutline-legend-3"
        style="width: 0.01px;"
      >
        <span>
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
